src/main/serialport.ts

```diff
@@ -0,0 +1,348 @@
+            throw new Error('PORT_IS_ALREADY_BEING_OPENED');
+        }
+
+        const renderers = existingPort.renderers;
+        const alreadyInList = renderers.find(
+            existing => sender.id === existing.id
+        );
+
+        if (!alreadyInList) {
+            let isDifferentSettings;
+            if (existingPort.serialPort) {
+                const currentOptions = existingPort.serialPort.settings;
+                Object.entries(options).forEach(([key, value]) => {
+                    if (
+                        !(key in currentOptions) ||
+                        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+                        (currentOptions as any)[key] !== value
+                    ) {
+                        isDifferentSettings = true;
+                    }
+                });
+            }
+
+            if (isDifferentSettings) {
+                if (!overwrite) {
+                    throw new Error('FAILED_DIFFERENT_SETTINGS');
+                }
+
+                const result = await changeOptions(options);
+                if (result === 'SUCCESS') {
+                    existingPort.renderers.forEach(renderer => {
+                        renderer.send(SERIALPORT_CHANNEL.ON_CHANGED, options);
+                    });
+                    renderers.push(sender);
+                    sender.on('destroyed', () => {
+                        console.log('Window Closed!');
+                        removeRenderer(path, sender);
+                    });
+                    return;
+                }
+                throw new Error('FAILED');
+            }
+            renderers.push(sender);
+            sender.on('destroyed', () => {
+                console.log('Window Closed!');
+                removeRenderer(path, sender);
+            });
+        }
+        return;
+    }
+
+    let port: SerialPort<AutoDetectTypes>;
+    try {
+        port = new SerialPort({ ...options, autoOpen: false });
+    } catch (error) {
+        throw new Error('FAILED');
+    }
+    const openPort: OpenPort = {
+        serialPort: port,
+        renderers: [sender],
+        settingsLocked: false,
+        opening: true,
+    };
+    serialPorts.set(path, openPort);
+
+    // Remove renderer if renderer process is destroyed (window is closed).
+    sender.on('destroyed', () => removeRenderer(path, sender));
+    // Remove renderer if renderer window is reloaded
+    sender.on('did-finish-load', () => {
+        removeRenderer(path, sender);
+    });
+
+    port.on('data', data => {
+        onData(path, data);
+    });
+
+    port.on('close', (error: Error) => {
+        // Device powers off, or is closed manually: do a clean up
+        if (error) {
+            if (openPort) {
+                openPort.renderers.forEach(renderer => {
+                    renderer.send(SERIALPORT_CHANNEL.ON_CLOSED);
+                });
+
+                serialPorts.delete(path);
+            }
+        }
```

Comment:
Question: what happens if port is Opening and an we disconnect do we still clean up on the opening exception?

Reply:
If I understand correctly, I'm not even able to produce this scenario:

> 2022-11-30T13:20:17.390Z INFO SerialPort: Port with path=/dev/ttyACM0 has been opened. 
> 2022-11-30T13:20:17.800Z INFO SerialPort: Port with path=/dev/ttyACM0 was asked if it's open, and returned true 
> 2022-11-30T13:20:17.803Z INFO SerialPort: Port with path=/dev/ttyACM0 have removed renderer with id=3 from its renderers list. 
> 2022-11-30T13:20:17.804Z INFO SerialPort: Port with path=/dev/ttyACM0 have an empty renderers list and will be closed. 
> 2022-11-30T13:20:17.808Z INFO SerialPort: Port with path=/dev/ttyACM0 was closed quietly